### PR TITLE
Openlr tool.  Partner's map on the side.

### PR DESCRIPTION
**Pull request by mgsergio**
@ygorshenin @igrechuhin

### CODE REVIEW

**Comment by igrechuhin** [comment]
LGTM

File: openlr/openlr_match_quality/openlr_assessment_tool/main.cpp [hunk 1]
```diff
@@ -6,11 +6,30 @@
 
 #include <QApplication>
 
+namespace
+{
 DEFINE_string(resources_path, "", "Path to resources directory");
 DEFINE_string(data_path, "", "Path to data directory");
+DEFINE_string(login, "", "Login string");
+DEFINE_string(paswd, "", "Password string");
```
Comment (ygorshenin) [inline openlr/openlr_match_quality/openlr_assessment_tool/main.cpp]: s/paswd/password/
No need to keep beautiful string length invariant here.

File: openlr/openlr_match_quality/openlr_assessment_tool/main.cpp [hunk 2]
```diff
@@ -6,11 +6,30 @@
...
+DEFINE_string(url, "", "Url to a partner map");
+
+bool ValidateStringFlag(char const * flagName, std::string const & val)
+{
+  if (!val.empty())
+    return true;
+
+  LOG(LERROR, (flagName, "cannot be empty. Please specify a proper", flagName));
```
Comment (ygorshenin) [inline openlr/openlr_match_quality/openlr_assessment_tool/main.cpp]: fprintf to stderr, please. Note! Logging system may not be initialized at this point, fprintf is much safer.

File: openlr/openlr_match_quality/openlr_assessment_tool/mainwindow.cpp [hunk 3]
```diff
@@ -226,49 +231,76 @@ class PointsControllerDelegate : public PointsControllerDelegateBase
 };
 }  // namespace
 
-
-MainWindow::MainWindow(Framework & framework)
+MainWindow::MainWindow(Framework & framework, std::string const & url, std::string const & login,
+                       std::string const & paswd)
```
Comment (ygorshenin) [inline openlr/openlr_match_quality/openlr_assessment_tool/mainwindow.cpp]: s/paswd/password/g

File: openlr/openlr_match_quality/openlr_assessment_tool/mainwindow.cpp [hunk 4]
```diff
@@ -226,49 +231,76 @@ class PointsControllerDelegate : public PointsControllerDelegateBase
...
   : m_framework(framework)
 {
+
```
Comment (ygorshenin) [inline openlr/openlr_match_quality/openlr_assessment_tool/mainwindow.cpp]: Seems that you don't need this blank line?

File: openlr/openlr_match_quality/openlr_assessment_tool/points_controller_delegate_base.hpp [hunk 5]
```diff
@@ -0,0 +1,33 @@
+#pragma once
+
+#include "indexer/feature.hpp"
+
+#include "geometry/point2d.hpp"
+
+#include <vector>
+
+using FeaturePoint = std::pair<FeatureID, size_t>;
```
Comment (ygorshenin) [inline openlr/openlr_match_quality/openlr_assessment_tool/points_controller_delegate_base.hpp]: ```#include <cstddef>``` for size_t.

File: openlr/openlr_match_quality/openlr_assessment_tool/points_controller_delegate_base.hpp [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    Add,
+    Remove
+  };
+
+  virtual std::vector<m2::PointD> GetAllJunctionPointsInViewport() const = 0;
+  /// Returns all junction points at a given location in the form of feature id and
+  /// point index in the feature.
+  virtual std::pair<std::vector<FeaturePoint>, m2::PointD> GetCandidatePoints(
+      m2::PointD const & p) const = 0;
+  virtual std::vector<m2::PointD> GetReachablePoints(m2::PointD const & p) const = 0;
```
Comment (ygorshenin) [inline openlr/openlr_match_quality/openlr_assessment_tool/points_controller_delegate_base.hpp]: Comment to the method, please.

File: openlr/openlr_match_quality/openlr_assessment_tool/segment_correspondence.hpp [hunk 7]
```diff
@@ -0,0 +1,57 @@
+#pragma once
+
+#include "openlr/decoded_path.hpp"
+#include "openlr/openlr_model.hpp"
+
+#include "3party/pugixml/src/pugixml.hpp"
+
+class SegmentCorrespondence
```
Comment (ygorshenin) [inline openlr/openlr_match_quality/openlr_assessment_tool/segment_correspondence.hpp]: Namespace?

File: openlr/openlr_match_quality/openlr_assessment_tool/traffic_drawer_delegate_base.hpp [hunk 8]
```diff
@@ -0,0 +1,22 @@
+#pragma once
+
+#include <geometry/point2d.hpp>
+
+/// This class is used to delegate segments drawing to the DrapeEngine.
```
Comment (ygorshenin) [inline openlr/openlr_match_quality/openlr_assessment_tool/traffic_drawer_delegate_base.hpp]: Put this into named namespace, please.

File: openlr/openlr_match_quality/openlr_assessment_tool/traffic_mode.cpp [hunk 9]
```diff
@@ -257,7 +258,7 @@ void TrafficMode::StartBuildingPath()
   if (m_buildingPath)
     MYTHROW(TrafficModeError, ("Path building already in progress."));
 
-  if (m_currentSegment->GetGoldenPath())
+  if (!m_currentSegment->GetGoldenPath().empty())
```
Comment (ygorshenin) [inline openlr/openlr_match_quality/openlr_assessment_tool/traffic_mode.cpp]: How about to implement short method named HasGoldenPath?

File: openlr/openlr_match_quality/openlr_assessment_tool/web_view.cpp [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+    }
+  });
+
+  connect(this, &QWebEngineView::renderProcessTerminated,
+          [this](QWebEnginePage::RenderProcessTerminationStatus termStatus, int statusCode) {
+            QString status;
+            switch (termStatus)
+            {
+            case QWebEnginePage::NormalTerminationStatus:
+              status = tr("Render process normal exit");
```
Comment (ygorshenin) [inline openlr/openlr_match_quality/openlr_assessment_tool/web_view.cpp]: Are you sure that we need translations here?

File: openlr/openlr_match_quality/openlr_assessment_tool/web_view.hpp [hunk 11]
```diff
@@ -0,0 +1,40 @@
+#pragma once
+
+#include <QWebEngineView>
+
+class WebView : public QWebEngineView
```
Comment (ygorshenin) [inline openlr/openlr_match_quality/openlr_assessment_tool/web_view.hpp]: Namespace, please...

File: openlr/openlr_match_quality/openlr_assessment_tool/web_view.hpp [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+private:
+  void GoToSegment();
+  void Login();
+
+  int m_loadProgress;
+  int m_requestedSegmentId = kInvalidId;
+
+  std::string m_url;
+  std::string m_login;
+  std::string m_paswd;
```
Comment (ygorshenin) [inline openlr/openlr_match_quality/openlr_assessment_tool/web_view.hpp]: m_password

File: openlr/openlr_match_quality/openlr_assessment_tool/web_view.cpp [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+
+void WebView::GoToSegment()
+{
+  CHECK(!IsLoginUrl(url().toString()), ("Can't go to a segment on the login page."));
+  if (m_requestedSegmentId == kInvalidId)
+    return;
+
+  auto const script = QString(R"EOT(
+    function s(ctx, arg) { return ctx.querySelector(arg); }
+    function turnOff(cb) { if (cb.checked) cb.click(); }
```
Comment (ygorshenin) [inline openlr/openlr_match_quality/openlr_assessment_tool/web_view.cpp]: uncheck?

File: openlr/openlr_match_quality/openlr_assessment_tool/web_view.cpp [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+}
+
+void WebView::GoToSegment()
+{
+  CHECK(!IsLoginUrl(url().toString()), ("Can't go to a segment on the login page."));
+  if (m_requestedSegmentId == kInvalidId)
+    return;
+
+  auto const script = QString(R"EOT(
+    function s(ctx, arg) { return ctx.querySelector(arg); }
```
Comment (ygorshenin) [inline openlr/openlr_match_quality/openlr_assessment_tool/web_view.cpp]: s/s/select/

File: openlr/openlr_match_quality/openlr_assessment_tool/main.cpp (hunk 1 above)
Comment (mgsergio) [reply openlr/openlr_match_quality/openlr_assessment_tool/main.cpp]: Done.

File: openlr/openlr_match_quality/openlr_assessment_tool/main.cpp (hunk 2 above)
Comment (mgsergio) [reply openlr/openlr_match_quality/openlr_assessment_tool/main.cpp]: Done.

File: openlr/openlr_match_quality/openlr_assessment_tool/mainwindow.cpp (hunk 3 above)
Comment (mgsergio) [reply openlr/openlr_match_quality/openlr_assessment_tool/mainwindow.cpp]: Done.

File: openlr/openlr_match_quality/openlr_assessment_tool/mainwindow.cpp (hunk 4 above)
Comment (mgsergio) [reply openlr/openlr_match_quality/openlr_assessment_tool/mainwindow.cpp]: Done.

File: openlr/openlr_match_quality/openlr_assessment_tool/points_controller_delegate_base.hpp (hunk 5 above)
Comment (mgsergio) [reply openlr/openlr_match_quality/openlr_assessment_tool/points_controller_delegate_base.hpp]: Done.

File: openlr/openlr_match_quality/openlr_assessment_tool/web_view.cpp (hunk 10 above)
Comment (mgsergio) [reply openlr/openlr_match_quality/openlr_assessment_tool/web_view.cpp]: Done.

File: openlr/openlr_match_quality/openlr_assessment_tool/web_view.hpp (hunk 12 above)
Comment (mgsergio) [reply openlr/openlr_match_quality/openlr_assessment_tool/web_view.hpp]: Done.

File: openlr/openlr_match_quality/openlr_assessment_tool/web_view.cpp (hunk 13 above)
Comment (mgsergio) [reply openlr/openlr_match_quality/openlr_assessment_tool/web_view.cpp]: Done.

File: openlr/openlr_match_quality/openlr_assessment_tool/web_view.cpp (hunk 14 above)
Comment (mgsergio) [reply openlr/openlr_match_quality/openlr_assessment_tool/web_view.cpp]: Done.

File: openlr/openlr_match_quality/openlr_assessment_tool/points_controller_delegate_base.hpp (hunk 6 above)
Comment (mgsergio) [reply openlr/openlr_match_quality/openlr_assessment_tool/points_controller_delegate_base.hpp]: Done.

File: openlr/openlr_match_quality/openlr_assessment_tool/segment_correspondence.hpp (hunk 7 above)
Comment (mgsergio) [reply openlr/openlr_match_quality/openlr_assessment_tool/segment_correspondence.hpp]: Done.

File: openlr/openlr_match_quality/openlr_assessment_tool/traffic_drawer_delegate_base.hpp (hunk 8 above)
Comment (mgsergio) [reply openlr/openlr_match_quality/openlr_assessment_tool/traffic_drawer_delegate_base.hpp]: Done.

File: openlr/openlr_match_quality/openlr_assessment_tool/web_view.hpp (hunk 11 above)
Comment (mgsergio) [reply openlr/openlr_match_quality/openlr_assessment_tool/web_view.hpp]: Done.

File: openlr/openlr_match_quality/openlr_assessment_tool/traffic_mode.cpp (hunk 9 above)
Comment (mgsergio) [reply openlr/openlr_match_quality/openlr_assessment_tool/traffic_mode.cpp]: Done.

**Comment by mgsergio** [comment]
PTAL

File: openlr/openlr_match_quality/openlr_assessment_tool/main.cpp [hunk 15]
```diff
@@ -4,13 +4,36 @@
...
+DEFINE_string(login, "", "Login string");
+DEFINE_string(password, "", "Password string");
+DEFINE_string(url, "", "Url to a partner map");
+
+bool ValidateStringFlag(char const * flagName, std::string const & val)
+{
+  if (!val.empty())
+    return true;
+
+  fprintf(stderr, "%s %s, %s", flagName, " cannot be empty. Please specify a proper ", flagName);
```
Comment (ygorshenin) [inline openlr/openlr_match_quality/openlr_assessment_tool/main.cpp]:
```c++
fprintf(stderr, "%s cannot be empty, Please, specify a proper %s\n", flagName, flagName);
```
?

Comment (mgsergio) [reply openlr/openlr_match_quality/openlr_assessment_tool/main.cpp]: Done.

**Comment by mgsergio** [comment]
PTAL

**Pull request merged**